paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/GlobalFullCompactionSinkWrite.java

```diff
@@ -0,0 +1,254 @@
+    }
+
+    @Override
+    public void compact(BinaryRow partition, int bucket, boolean fullCompaction) throws Exception {
+        super.compact(partition, bucket, fullCompaction);
+        touchBucket(partition, bucket);
+    }
+
+    private void touchBucket(BinaryRow partition, int bucket) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("touch partition {}, bucket {}", partition, bucket);
+        }
+
+        // partition is a reused BinaryRow
+        // we first check if the tuple exists to minimize copying
+        if (!currentWrittenBuckets.contains(Tuple2.of(partition, bucket))) {
+            currentWrittenBuckets.add(Tuple2.of(partition.copy(), bucket));
+        }
+    }
+
+    @Override
+    public List<Committable> prepareCommit(boolean doCompaction, long checkpointId)
+            throws IOException {
+        checkSuccessfulFullCompaction();
+
+        // collects what buckets we've modified during this checkpoint interval
+        if (!currentWrittenBuckets.isEmpty()) {
+            writtenBuckets
+                    .computeIfAbsent(checkpointId, k -> new HashSet<>())
+                    .addAll(currentWrittenBuckets);
+            currentWrittenBuckets.clear();
+        }
+
+        if (LOG.isDebugEnabled()) {
+            for (Map.Entry<Long, Set<Tuple2<BinaryRow, Integer>>> checkpointIdAndBuckets :
+                    writtenBuckets.entrySet()) {
+                LOG.debug(
+                        "Written buckets for checkpoint #{} are:", checkpointIdAndBuckets.getKey());
+                for (Tuple2<BinaryRow, Integer> bucket : checkpointIdAndBuckets.getValue()) {
+                    LOG.debug("  * partition {}, bucket {}", bucket.f0, bucket.f1);
+                }
+            }
+        }
+
+        if (!writtenBuckets.isEmpty() && isFullCompactedIdentifier(checkpointId, deltaCommits)) {
+            doCompaction = true;
+        }
+
+        if (doCompaction) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Submit full compaction for checkpoint #{}", checkpointId);
+            }
+            submitFullCompaction(checkpointId);
+            commitIdentifiersToCheck.add(checkpointId);
+        }
+
+        return super.prepareCommit(doCompaction, checkpointId);
+    }
+
+    private void checkSuccessfulFullCompaction() {
+        Long latestId = snapshotManager.latestSnapshotId();
+        if (latestId == null) {
+            return;
+        }
+        Long earliestId = snapshotManager.earliestSnapshotId();
+        if (earliestId == null) {
+            return;
+        }
+
+        for (long id = latestId; id >= earliestId; id--) {
+            Snapshot snapshot = snapshotManager.snapshot(id);
+            if (snapshot.commitUser().equals(commitUser)
+                    && snapshot.commitKind() == Snapshot.CommitKind.COMPACT) {
+                long commitIdentifier = snapshot.commitIdentifier();
+                if (commitIdentifiersToCheck.contains(commitIdentifier)) {
```

Comment:
Is `commitIdentifiersToCheck` only used for log purpose?

Reply:
Take a look to `checkSuccessfulFullCompaction`.